Control-User-Interfaces/writing-sal-scripts.rst

```diff
@@ -0,0 +1,948 @@
+
+.. code-block:: python
+
+  def set_metadata(self, metadata):
+    """Set estimated duration of the script.
+    """
+    # Add 30s for slew
+    estimated_slew_time = 30.
+
+    # Estimate time taking data; basically sum the exposure times
+    data_duration = np.sum(self.config.exptime)
+
+    metadata.duration = estimated_slew_time + data_duration
+
+.. _Writing-SAL-Scripts-The-run-Method:
+
+The ``run`` Method
+------------------
+
+The ``run`` method in the SAL Script is responsible for executing the operation, it is called once the script "transitions" to "running" state.
+
+Having developed the procedure in the notebook already, we can simply transfer the code to this method.
+We will also have perform some minor modifications to use the configuration parameters.
+
+Furthermore, we may also want to add some ``checkpoints`` to the process.
+These are calls done in the ``run`` method (you can not use them in other places of the code) of a SAL Script where it is possible to pause the execution.
+
+Taking all that into consideration, this is what the ``run`` method of this SAL Script would look like:
+
+.. code-block:: python
+  :linenos:
+
+  async def run(self):
+
+    self.log.debug(
+      f"Slew to {self.config.target_name} with "
+      f"rot_value: {self.config.rot_value} and rot_type: {self.config.rot_type}"
+    )
+
+    await atcs.slew_object(
+        name=self.config.target_name,
+        rot=self.config.rot_value,
+        rot_type=self.config.rot_type
+      )
+
+    # After slew is a good idea to add a checkpoint.
+    await self.checkpoint("Slew finished")
+
+    # Compute grid
+    n_grid = len(self.config.exptime)
+    grid_x = (np.random.rand(n_grid)-0.5)*120.
+    grid_y = (np.random.rand(n_grid)-0.5)*120.
+
+    for iter, xx, yy in zip(range(n_grid), grid_x, grid_y):
+
+      # Let's add some debugging messages
+      self.log.debug(
+        f"Starting iteration {iter+1} of {n_grid}: "
+        f"offset x/y: {xx},{yy} arcsec."
+      )
+
+      # We may also want to add a checkpoint at the start of each iteration
+      await self.checkpoint(f"iter[{iter+1}/{n_grid}]")
+
+      # Offset telescope
+      # Use non-relative offset as they are easier to reset
+      await atcs.offset_xy(x=xx, y=yy, relative=False)
+
+      # Take data
+      for etime, flt, grt in zip(
+          self.config.exptime, self.config.obs_filter, self.config.obs_grating
+      ):
+        # We can also add checkpoints at before every image, this will give us
+        # more granularity in controlling the script.
+        await self.checkpoint(
+          f"iter[{iter+1}/{n_grid}]: "
+          f"exptime: {etime}, filter: {flt}, grating: {grt}"
+        )
+        await latiss.take_object(exptime=etime, filter=flt, grating=grt)
+
+    # Reset offset
+    await atcs.offset_xy(x=0., y=0., relative=False)
+
+Once this is completed you are pretty much done.
+The full source for the SAL Script worked in this example can be found here.
```

Comment:
what about using `arun` versus `run` such that the script doesn't close out all the remotes when it's finished?

Reply:
I am not sure if we want to get into this level of detail here. It is not necessarily part of the SAL Script base interface. Maybe mention it in an appendix like section?